### PR TITLE
GBM: support hyperopt

### PR DESCRIPTION
**Pull request by jppgks**
Runs evaluation after every `boosting_rounds_per_checkpoint` to populate training/validation metrics as required by hyperopt

Mimics as much as possible the ECD train loop logic

### CODE REVIEW

File: ludwig/trainers/trainer_lightgbm.py [hunk 1]
```diff
@@ -319,51 +308,149 @@ def run_evaluation(
...
 
-        for epoch, steps in enumerate(range(0, self.num_boost_round, self.boosting_round_log_frequency), start=1):
-            progress_tracker.epoch = epoch
+        return should_break
 
-            evals_result = {}
-            booster = self.train_step(
-                params, lgb_train, eval_sets, eval_names, booster, self.boosting_round_log_frequency, evals_result
-            )
+    def check_progress_on_validation(
```
Comment (justinxzhao) [inline ludwig/trainers/trainer_lightgbm.py:392]: Is this function the same as the ECD trainer? If so, we should factor this out, like you've done with append_metrics.

Comment (jppgks) [reply ludwig/trainers/trainer_lightgbm.py:392]: Not entirely, ECD trainer has additional logic here for reducing learning rate and increasing batch size

**Review by justinxzhao — APPROVED** [review]
